Piwik.Tracker.Tests/PiwikTrackerWithMockedServerTests.cs

```diff
@@ -0,0 +1,516 @@
+        private const string UA = "Firefox";
+        private const string PiwikBaseUrl = "http://127.0.0.1:1122/piwik.php";
+        private const int SiteId = 1;
+
+        private static readonly NameValueCollection DefaultRequestParameter = new NameValueCollection
+        {
+            { "idsite",SiteId.ToString()},
+            { "rec","1"},
+            { "apiv","1"},
+            { "url","http://unknown" },
+        };
+
+        private static readonly string[] DefaultRequestParameterKeysToRemoveFromComparison =
+        {
+            "r", // random value
+            "_idts" // _createTs from cookie
+        };
+
+        [OneTimeSetUp]
+        public void SetUpFixture()
+        {
+            _mockedPiwikServer = new MockedHttpServer(PiwikBaseUrl);
+        }
+
+        [OneTimeTearDown]
+        public void TearDownFixture()
+        {
+            _mockedPiwikServer.Dispose();
+        }
+
+        [SetUp]
+        public void SetUpTest()
+        {
+            _sut = new PiwikTracker(SiteId, PiwikBaseUrl);
+        }
+
+        [Test]
+        [TestCase("myPage")]
+        [TestCase("myPage/?Ü&")]
+        public void DoTrackPageView_Test(string documentTitle)
+        {
+            //Arrange
+            var retrieveRequest = CreateAllGetOkRequestBehavior();
+            //Act
+            var actual = _sut.DoTrackPageView(documentTitle);
+            //Assert
+            var actualRequest = retrieveRequest();
+            Assert.NotNull(actualRequest);
+            Assert.That(actual.HttpStatusCode, Is.EqualTo(HttpStatusCode.OK));
+            Assert.That(actualRequest.Method, Is.EqualTo("GET"));
+            Assert.That(actualRequest.RequestUri.GetAuthorityAndPath(), Is.EqualTo(PiwikBaseUrl));
+            Assert.PiwikRequestParameterMatch(actualRequest.RequestUri.ParseQueryString(), new NameValueCollection
+            {
+                { "_idvc", "0" },// visit count
+                { "_id", _sut.GetVisitorId() }, // visitor id
+                { "action_name", documentTitle },
+            });
+        }
+
+        [Test]
+        [TestCase("myCategory", "myAction", "myName", "myValue")]
+        [TestCase("myCategory", "myAction", "myName", "")]
+        [TestCase("myCategory", "myAction", "", "myValue")]
+        public void DoTrackEvent_Test(string category, string action, string name, string value)
+        {
+            //Arrange
+            var retrieveRequest = CreateAllGetOkRequestBehavior();
+            //Act
+            var actual = _sut.DoTrackEvent(category, action, name, value);
+            //Assert
+            var actualRequest = retrieveRequest();
+            Assert.NotNull(actualRequest);
+            Assert.That(actual.HttpStatusCode, Is.EqualTo(HttpStatusCode.OK));
+            Assert.That(actualRequest.Method, Is.EqualTo("GET"));
+            Assert.That(actualRequest.RequestUri.GetAuthorityAndPath(), Is.EqualTo(PiwikBaseUrl));
+            if (string.IsNullOrEmpty(name))
```

Comment:
Would it be possible to reduce the amount of duplicated code?

Wouldn't it be possible to do something like : 

```
parameters = new NameValueCollection
                 {
                     { "_idvc", "0" },// visit count
                     { "_id", _sut.GetVisitorId() }, // visitor id
                     { "e_c", category },
                 }

if (!string.IsNullOrEmpty(name))
    parameters.Add({ "e_n", name })

// and so on..
```

Wouldn't this help with the fact that common parameters are currently duplicated?

If this approach would be applied to the other methods we may see a substantial amount of lines removed.